File: implementation/Scala/src/main/java/synapses/jvm/library/NeuralNetwork.java
package synapses.jvm.library;

import scala.collection.immutable.LazyList;
import synapses.JavaInterface.NeuralNetwork$;

import java.util.function.IntFunction;

public class NeuralNetwork {

    LazyList contents;

    public NeuralNetwork(LazyList _contents) {
        contents = _contents;
    }

    public static NeuralNetwork init(int[] layers) {
        LazyList _contents = NeuralNetwork$.MODULE$.init(layers);
        return new NeuralNetwork(_contents);
    }

    public static NeuralNetwork initWithSeed(int seed, int[] layers) {
        LazyList _contents = NeuralNetwork$.MODULE$.initWithSeed(seed, layers);
        return new NeuralNetwork(_contents);
    }

    public static NeuralNetwork customizedInit(
            int[] layers,
            IntFunction<ActivationFunction> activationF,
            IntFunction<Double> weightInitF) {
        LazyList _contents = NeuralNetwork$.MODULE$.customizedInit(
                layers,
                (i) -> activationF.apply(i).contents,
                weightInitF::apply
        );
        return new NeuralNetwork(_contents);
    }

    public static double[] prediction(NeuralNetwork network, double[] inputValues){
        return NeuralNetwork$.MODULE$.prediction(network.contents, inputValues);
    }

    public static double[] errors(
            NeuralNetwork network,
            double learningRate,
            double[] inputValues,
            double[] expectedOutput){
        return NeuralNetwork$.MODULE$.errors(network.contents, learningRate, inputValues, expectedOutput);
    }

    public static NeuralNetwork fit(
            NeuralNetwork network,
            double learningRate,
            double[] inputValues,
            double[] expectedOutput){
        LazyList _contents = NeuralNetwork$.MODULE$.fit(network.contents, learningRate, inputValues, expectedOutput);
        return new NeuralNetwork(_contents);
    }

    public static String toJson(NeuralNetwork network){
        return NeuralNetwork$.MODULE$.toJson(network.contents);
    }

    public static NeuralNetwork ofJson(String json){
        LazyList _contents = NeuralNetwork$.MODULE$.ofJson(json);
        return new NeuralNetwork(_contents);
    }

    public static String toSvg(NeuralNetwork network){
        return NeuralNetwork$.MODULE$.toSvg(network.contents);
    }

}

File: implementation/DotNet/SynapsesCSharp/NeuralNetwork.cs
using System;
using System.Linq;
using FSharpx.Collections;
using Microsoft.FSharp.Collections;
using Microsoft.FSharp.Core;
using Synapses;
using Synapses.Model;
using Synapses.Model.NetElems;

namespace SynapsesCSharp
{
    public class NeuralNetwork
    {
        readonly LazyList<LazyList<Neuron.Neuron>> contents;

        public NeuralNetwork(LazyList<LazyList<Neuron.Neuron>> _contents)
        {
            contents = _contents;
        }

        public static NeuralNetwork init(int[] layers)
        {
            FSharpList<int> ls = ListModule.OfSeq(layers);
            LazyList<LazyList<Neuron.Neuron>> _contents =
                NeuralNetworkModule.init(ls);
            return new NeuralNetwork(_contents);
        }

        public static NeuralNetwork initWithSeed(int seed, int[] layers)
        {
            FSharpList<int> ls = ListModule.OfSeq(layers);
            LazyList<LazyList<Neuron.Neuron>> _contents =
                NeuralNetworkModule.initWithSeed(seed, ls);
            return new NeuralNetwork(_contents);
        }

        public static NeuralNetwork customizedInit(
            int[] layers,
            Func<int, ActivationFunction> activationF,
            Func<int, Double> weightInitF)
        {
            FSharpList<int> ls = ListModule.OfSeq(layers);
            FSharpFunc<int, Activation.Function> actF =
                (Converter<int, Activation.Function>) (x => activationF(x).contents);
            FSharpFunc<int, Double> weightF =
                (Converter<int, Double>) (x => weightInitF(x));
            LazyList<LazyList<Neuron.Neuron>> _contents =
                NeuralNetworkModule.customizedInit(ls, actF, weightF);
            return new NeuralNetwork(_contents);
        }
        
        public static double[] prediction(NeuralNetwork network, double[] inputValues){
            FSharpList<double> ivs = ListModule.OfSeq(inputValues);
            return NeuralNetworkModule
                .prediction(network.contents, ivs)
                .ToArray();
        }
        
        public static double[] errors(
            NeuralNetwork network,
            double learningRate,
            double[] inputValues,
            double[] expectedOutput){
            FSharpList<double> ivs = ListModule.OfSeq(inputValues);
            FSharpList<double> eos = ListModule.OfSeq(expectedOutput);
            return NeuralNetworkModule
                .errors(network.contents, learningRate, ivs, eos)
                .ToArray();
        }
        
        public static NeuralNetwork fit(
            NeuralNetwork network,
            double learningRate,
            double[] inputValues,
            double[] expectedOutput){
            FSharpList<double> ivs = ListModule.OfSeq(inputValues);
            FSharpList<double> eos = ListModule.OfSeq(expectedOutput);
            LazyList<LazyList<Neuron.Neuron>> _contents =
                NeuralNetworkModule.fit(network.contents, learningRate, ivs, eos);
            return new NeuralNetwork(_contents);
        }
        
        public static string toJson(NeuralNetwork network){
            return NeuralNetworkModule.toJson(network.contents);
        }
        
        public static NeuralNetwork ofJson(string json){
            LazyList<LazyList<Neuron.Neuron>> _contents = NeuralNetworkModule.ofJson(json);
            return new NeuralNetwork(_contents);
        }
        
        public static string toSvg(NeuralNetwork network){
            return NeuralNetworkModule.toSvg(network.contents);
        }
        
    }
}